### PR TITLE
Enhance the bound generator concept to remove duplicate parameters and make it more pervasive throughout the system

### LINKED ISSUE #89033: Marshalling source generation code could pass TypePositionInfo and StubCodeContext only to constructors and have parameterless generation methods
The marshalling generator types in Microsoft.Interop.SourceGeneration pass a StubCodeContext to a method that builds an IMarshallingGenerator, which takes a TypePositionInfo and StubCodeContext each time it generates a method, and has different behavior depending on those. 

If instead, a constructor method composed an object based on the StubCodeContext and TypePositionInfo, and the methods on the object took no parameters, the generator could be much easier to debug and unit test, as the composition of the object would be the only thing that impacts the generated code. This would also make it easy to unit test, since tests could compose an object and test the behavior of the generator without need to create a TypePositionInfo and StubCodeContext.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
This reduces the number of places where we need to manually pass around a marshalling generator and its matching `TypePositionInfo`.

This PR also simplifies some portions of the JS generators.

Addresses https://github.com/dotnet/runtime/issues/89033

Moving the generators to also store the stub code context would end up requiring us to rebind in the majority of cases, increasing GC load and not actually improving maintainability much.

### CODE REVIEW

**Comment by pavelsavara** [comment]
It would be great to introduce some test which would assert the explicit shape of the output code, so that we feel safer doing refactorings.

We had situation that the previous refactoring introduced [duplicate statement](https://github.com/dotnet/runtime/issues/92400) in the generated code and we have not noticed for quite long time. [more details](https://github.com/dotnet/runtime/pull/92403)

**Comment by jkoritzinsky** [comment]
I've added a test that validates the exact code output (and fixed a case of non-determinism in the JS generators in the process)

**Review by jtschuster — APPROVED** [review]
Looks good to me, but since this is a big change and just a refactor, we should probably wait until after we snap for 9 to merge

**Comment by jkoritzinsky** [comment]
I'll move this and the corresponding issue to .NET 10.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:88]: I'm assume this is what the other source generators do? I assume we can all agree whatever one does, they are should do with respect to line endings.

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:88]: Other generators generate strings with the system line endings. However, since we use Roslyn, we get their line endings support (which defaults to CRLF). And we can't use Environment.NewLine as Roslyn bans the Environment class in analyzers.

We're only hitting this now because we're trying to use the Roslyn testing SDK's support for comparing against known output, which is line-ending sensitive.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**
